[Workers] WASM Rust Guide (#8539)
cloudflare/cloudflare-docs: content/workers/platform/web-assembly/rust.md

---
title: Rust
pcx_content_type: rust-wasm-guide
weight: 1
meta:

>   title: Rust WebAssembly guide

Reviewer: @zebp may have some thoughts on this

Zeb — goal here is to have place to point people to that helps people go from zero to one, even if they're otherwise not familiar with Workers, wrangler, etc.

---

# Rust WebAssembly Guide

> This guide will show how to get started with using Rust to build WebAssembly Workers. There are many configurations that can be used to produce Wasm binaries which are compatible with Workers, but this guide will focus on using [`workers-rs`](https://github.com/cloudflare/workers-rs), which provides the most fully-featured experience. 

Reviewer: ```suggestion
This guide will show how to get started with using Rust to build WebAssembly Workers. There are many configurations that can be used to produce Wasm binaries which are compatible with Workers, but this guide will focus on using [`workers-rs`](https://github.com/cloudflare/workers-rs), which makes APIs and bindings like Workers KV, R2, and Queues available directly from Rust. 
```

> This guide assumes that you have the following installed:

Reviewer: ```suggestion
## Prerequisites
```

Think @dcpena has some good docs patterns here

* A recent version of [`Rust`](https://rustup.rs/) 
* [`npm`](https://docs.npmjs.com/getting-started)
* The Rust `wasm32-unknown-unknown` toolchain:

```sh
$ rustup target add wasm32-unknown-unknown
```

## 1. Create a new project with Wrangler

Use `wrangler generate` to create a new project from Cloudflare's `workers-rs` template.

```sh

> $ npx wrangler generate hello-world-rust https://github.com/cloudflare/workers-sdk/templates/experimental/worker-rust

Reviewer: @lrapoport-cf anything changing here that we should be aware of when putting together docs?

Reviewer 2: @irvinebroque do you mean with the template or the command?

Reviewer: Command, I think I misunderstood what was changing, nvm

---
title: Rust
pcx_content_type: rust-wasm-guide
weight: 2
meta:
  title: Rust WebAssembly guide
---

# Rust WebAssembly Guide

> This guide demonstrates how to build a Worker entirely in the Rust programming language, using the `workers-rs` crate, which which makes [Runtime APIs](/workers/runtime-apis) and [bindings](/workers/platform/bindings/) like Workers KV, R2, and Queues available directly from your Rust code.

Author: ```suggestion
This guide demonstrates how to build a Worker entirely in the Rust programming language, using the `workers-rs` crate, which makes [Runtime APIs](/workers/runtime-apis) and [bindings](/workers/platform/bindings/) like Workers KV, R2, and Queues available directly from your Rust code.
```

```sh
$ rustup target add wasm32-unknown-unknown
```

## 1. Create a new project with Wrangler

Use `wrangler generate` to create a new project from Cloudflare's `workers-rs` template.

```sh
$ npx wrangler generate hello-world-rust https://github.com/cloudflare/workers-sdk/templates/experimental/worker-rust
```

Your project will be created in a new directory (`hello-world-rust`), to view its files:

```sh
$ cd hello-world-rust
```

Here you will find the following files and folders:

* `Cargo.toml` - The standard project configuration file for Rust's [`Cargo`](https://doc.rust-lang.org/cargo/) package manager. The template pre-populates some best-practice settings for building for Wasm on Workers.
* `README.md` - Boilerplate readme for working with the template project.
* `package.json` - NPM configuration for the template project which specifies useful commands (`dev` and `deploy`), and [Wrangler](https://github.com/cloudflare/workers-sdk/tree/main/packages/wrangler) as a dev-dependency.
* `wrangler.toml` - Wrangler configuration, pre-populated with a custom build command to invoke `worker-build` (See [bundling](/workers/platform/web-assembly/rust/#bundling-worker-build)).

> * `src` - Rust source directory, pre-populated with simple "Hello World" Worker.

Author: I don't think this is right, first, `/src` implies that it is at the root of their file system. If you feel that it is unclear that it is a directory, we can use `./src`. Second, I think Hello World should not be a code block.

* [`ObjectNamespace`](https://docs.rs/worker/latest/worker/durable/struct.ObjectNamespace.html) - [Durable Object](/workers/runtime-apis/durable-objects/#durable-objects) namespace binding.
* [`Fetcher`](https://docs.rs/worker/latest/worker/struct.Fetcher.html) - [Service Binding](/workers/runtime-apis/service-bindings/) to another Worker.
* [`Bucket`](https://docs.rs/worker/latest/worker/struct.Bucket.html) - [R2](/r2/) Bucket binding.

**[`Context`](https://docs.rs/worker/latest/worker/struct.Context.html)**

Provides access to [`waitUntil`](/workers/runtime-apis/fetch-event/#waituntil) (deferred asynchronous tasks) and [`passThroughOnException`](/workers/runtime-apis/fetch-event/#passthroughonexception) (fail open) functionality.

#### [`Response`](https://docs.rs/worker/latest/worker/struct.Response.html)

The `fetch` handler expects a `Response` return type, which includes support for streaming responses to the client asynchronously. This is also the return type of any subrequests made from your Worker. There
are methods for accessing status code and headers, as well as streaming the body asynchronously or deserializing from JSON using
[Serde](https://serde.rs/).

> #### [`Router`](https://docs.rs/worker/latest/worker/struct.Router.html)
> 
> Implements convenient routing API to serve multiple paths from one Worker, see [example](https://github.com/cloudflare/workers-rs#or-use-the-router).

Reviewer 3: ```suggestion
#### `Router`

Implements convenient [routing API](https://docs.rs/worker/latest/worker/struct.Router.html) to serve multiple paths from one Worker. Refer to the [`Router` example in the `worker-rs` GitHub repository](https://github.com/cloudflare/workers-rs#or-use-the-router).
```

> ## 4. Publish

Reviewer 3: ```suggestion
## 4. Publish your Worker project
```

> With your project configured, you can now publish your Worker, to a `*.workers.dev` subdomain, or a [custom domain](/workers/platform/triggers/custom-domains/), if you have one configured. If you have not configured any subdomain or domain, Wrangler will prompt you during the publish process to set one up.

Reviewer 3: ```suggestion
With your project configured, deploy your Worker by running the following command. You can publish your Worker to a `*.workers.dev` subdomain or a [custom domain](/workers/platform/triggers/custom-domains/), if you have one configured.
```
The last sentence is now true anymore in the Wrangler publish flow. I will update the Get started accordingly. 

`publish` is also being deprecated during Developer Week and this guide will need to be updated. @kflansburg @irvinebroque

```sh
$ npx wrangler publish
```

Preview your Worker at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev`.

> {{<Aside type="note" header="Note">}}

Reviewer 3: ```suggestion
{{<Aside type="note">}}
```

When pushing to your `*.workers.dev` subdomain for the first time, you may see [`523` errors](https://support.cloudflare.com/hc/articles/115003011431#523error) while DNS is propagating. These errors should resolve themselves after a minute or so.

{{</Aside>}}

> ## How This Works

Reviewer 3: ```suggestion
## How this deployment works
```

Wasm Workers are invoked from a JavaScript entrypoint script which is created automatically for you when using `workers-rs`. 

### JavaScript Plumbing (`wasm-bindgen`)

> To access platform features such as bindings, Wasm Workers must be able to access methods from the JavaScript runtime API. 
> This interoperability is achieved using [`wasm-bindgen`](https://rustwasm.github.io/wasm-bindgen/), which provides the glue code needed to import runtime APIs to, and export event handlers from, the Wasm module. `wasm-bindgen` also provides [`js-sys`](https://docs.rs/js-sys/latest/js_sys/),
> which implements types for interacting with JavaScript objects. In practice, this is an implementation detail, as `workers-rs`'s API handles conversion to and from JavaScript objects, and interaction with imported JavaScript runtime APIs for you.

Reviewer 3: ```suggestion
To access platform features such as bindings, Wasm Workers must be able to access methods from the JavaScript runtime API. 

This interoperability is achieved using [`wasm-bindgen`](https://rustwasm.github.io/wasm-bindgen/), which provides the glue code needed to import runtime APIs to, and export event handlers from, the Wasm module. `wasm-bindgen` also provides [`js-sys`](https://docs.rs/js-sys/latest/js_sys/), which implements types for interacting with JavaScript objects. In practice, this is an implementation detail, as `workers-rs`'s API handles conversion to and from JavaScript objects, and interaction with imported JavaScript runtime APIs for you.
```

### Async (`wasm-bindgen-futures`)

> [`wasm-bindgen-futures`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/) (part of the `wasm-bindgen` project) provides interoperability between Rust
> Futures and JavaScript Promises. `workers-rs` invokes the entire event handler function using `spawn_local`, meaning that you can program using async Rust, which is turned
> into a single JavaScript Promise and run on the JavaScript event loop. Calls to imported JavaScript runtime APIs are automatically converted to Rust Futures that can be easily invoked from async Rust functions.

Reviewer 3: ```suggestion
[`wasm-bindgen-futures`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/) (part of the `wasm-bindgen` project) provides interoperability between Rust Futures and JavaScript Promises. `workers-rs` invokes the entire event handler function using `spawn_local`, meaning that you can program using async Rust, which is turned into a single JavaScript `Promise` and run on the JavaScript event loop. Calls to imported JavaScript runtime APIs are automatically converted to Rust Futures that can be invoked from async Rust functions.
```

Author: Is this just to move to a single line? I believe the markdown will be formatted into a single paragraph, and lines this long are unpleasant to edit in a terminal.

### Bundling (`worker-build`)

To run the resulting Wasm binary on Workers, `workers-rs` includes a build tool called [`worker-build`](https://github.com/cloudflare/workers-rs/tree/main/worker-build) which: 

1. Creates a JavaScript entrypoint script that properly invokes the module using `wasm-bindgen`'s JavaScript API.  

> 2. Invokes `web-pack` to minify and bundle the JavaScript code.

Reviewer 4: We use `wasm-pack` to (which itself invokes `wasm-bindgen`'s cli) to generate the JS glue and minify the WASM bundle and we use `esbuild` to bundle and minify the JavaScript.

> 3. Outputs a directory structure that Wrangler can use to bundle and publish the final worker.

Reviewer 3: ```suggestion
3. Outputs a directory structure that Wrangler can use to bundle and publish the final Worker.
```
